lib/sigstore/verifier.rb

```diff
@@ -69,22 +86,41 @@ def verify(materials:, policy:)
         store.add_cert(cert.openssl)
       end
 
-      sign_date = materials.certificate.not_before
-
-      store.time = sign_date
-
-      store_ctx = OpenSSL::X509::StoreContext.new(store, materials.certificate.openssl)
+      # 3)
+      # The Verifier MUST perform certification path validation (RFC 5280 §6) of the certificate chain with the
+      # pre-distributed Fulcio root certificate(s) as a trust anchor, but with a fake “current time.”
+      # If a timestamp from the timestamping service is available, the Verifier MUST perform path validation using the
+      # timestamp from the Timestamping Service. If a timestamp from the Transparency Service is available, the Verifier
+      # MUST perform path validation using the timestamp from the Transparency Service. If both are available, the
+      # Verifier performs path validation twice. If either fails, verification fails.
+      chains = timestamps.map do |ts|
+        store_ctx = OpenSSL::X509::StoreContext.new(store, materials.certificate.openssl)
+        store_ctx.time = ts
+
+        unless store_ctx.verify
+          return VerificationFailure.new(
+            "failed to validate certification from fulcio cert chain: #{store_ctx.error_string}"
+          )
+        end
 
-      unless store_ctx.verify
-        return VerificationFailure.new(
-          "failed to validate certification from fulcio cert chain: #{store_ctx.error_string}"
-        )
+        chain = store_ctx.chain || raise(Error::InvalidCertificate, "no valid cert chain found")
+        chain.shift # remove the cert itself
+        chain.map! { Internal::X509::Certificate.new(_1) }
       end
 
-      chain = store_ctx.chain || raise(Error::InvalidCertificate, "no valid cert chain found")
-      chain.shift # remove the cert itself
-      chain.map! { Internal::X509::Certificate.new(_1) }
+      chains.uniq! { |chain| chain.map(&:to_der) }
+      unless chains.size == 1
```

Comment:
NB: Technically more than one valid chain is perfectly fine; X.509 validation semantics say that any satisfying chain can be chosen when this happens.

(I can't think of a reason why this would happen in this case, since at least with the Sigstore public instance there should really only be one chain. But conceptually there's nothing stopping it from happening.)

Reply:
yeah i was very iffy on this one. my bias is to have the stricter check, and refactor to validate SCTs against multiple chains if anyone actually hits this?

Reply:
Yeah, in theory you could have different chains for different timestamps. But agreed that the stricter check is probably fine for now 🙂.